mailserver/mailserver_db_leveldb.go

```diff
@@ -68,7 +69,35 @@ func NewLevelDB(dataDir string) (*LevelDB, error) {
 		log.Info("database is corrupted trying to recover", "path", dataDir)
 		db, err = leveldb.RecoverFile(dataDir, nil)
 	}
-	return &LevelDB{ldb: db}, err
+	instance := LevelDB{ldb: db}
+
+	// checking count on every insert is inefficient
+	go func() {
+		for {
+			if instance.closed {
```

Comment:
Exiting from a goroutine should be done using channels, usually:
```go
for {
    select {
    case <-instance.done:
        return
    case <-time.After(time.Second * envelopeCounterCheckInterval):
        instance.updateArchivedEnvelopesCounter()
    }
}
```

where `close` is `close chan struct{}` initialized in the constructor `instance: make(chan struct{})`.

Reply:
I think making the loop depend on the state of the DB is better, and I see no problem with checking DB state by checking type of error. But both LevelDB and SQL library don't expose properly those things. So I might as well change it as you described, though I don't think it's any better.